judge/views/problem.py

```diff
@@ -548,25 +548,33 @@ def problem_submit(request, problem=None, submission=None):
                                        _('You have been declared persona non grata for this problem. '
                                          'You are permanently barred from submitting this problem.'))
 
-            if profile.current_contest is not None:
-                try:
-                    contest_problem = form.cleaned_data['problem'].contests.get(contest=profile.current_contest.contest)
-                except ContestProblem.DoesNotExist:
-                    model = form.save()
+            with transaction.atomic():
+                if profile.current_contest is not None:
+                    try:
+                        contest_problem = form.cleaned_data['problem'].contests.get(contest=profile.current_contest.contest)
+                    except ContestProblem.DoesNotExist:
+                        model = form.save()
+                    else:
+                        max_subs = contest_problem.max_submissions
+                        if max_subs and get_contest_submission_count(problem, profile) >= max_subs:
+                            return generic_message(request, _('Too many submissions'),
+                                                    _('You have exceeded the submission limit for this problem.'))
+                        model = form.save()
+                        contest = ContestSubmission(submission=model, problem=contest_problem,
+                                                    participation=profile.current_contest)
+                        contest.save()
                 else:
-                    max_subs = contest_problem.max_submissions
-                    if max_subs and get_contest_submission_count(problem, profile) >= max_subs:
-                        return generic_message(request, _('Too many submissions'),
-                                                _('You have exceeded the submission limit for this problem.'))
                     model = form.save()
-                    contest = ContestSubmission(submission=model, problem=contest_problem,
-                                                participation=profile.current_contest)
-                    contest.save()
-            else:
-                model = form.save()
 
-            profile.update_contest()
-            model.judge(rejudge=False)
+                # Create the SubmissionSource object
+                source = SubmissionSource(submission=model, source=form.cleaned_data['source'])
+                source.save()
+                profile.update_contest()
+
+                # Save a query
+                model.source = source
```

Comment:
No, don't bother doing this inside the transaction. If the submission fails to be judged, it should still be saved.